### PR TITLE
[WIP] Adding WebWorker for the NFT markers

### PR DESCRIPTION
**Pull request by kalwalt**
Trying to add a WebWorker for better NFT performances.

### CODE REVIEW

**Comment by kalwalt** [comment]
@ThorstenBux netlify create also the preview! This PR is still a WIP but i hope to finish soon...

**Comment by ThorstenBux** [comment]
Yes, auto deploy of PRs. Awesome.

I get a 404 when I run the NFT tracking:
```
webWorker.js:42 GET https://5e98a11b28c3d40006ae43b8--sad-northcutt-73843d.netlify.app/editor/js/modules/artoolkit.worker.js 404
load @ webWorker.js:42
startWorker @ webWorker.js:141
(anonymous) @ arcontroller.component.js?version=0.64:168
(anonymous) @ arcontroller.component.js?version=0.64:158
(anonymous) @ artoolkit.min.js:1
writeCallback @ artoolkit.min.js:1
writeByteArrayToFS @ artoolkit.min.js:1
oReq.onload @ artoolkit.min.js:1
load (async)
ajax @ artoolkit.min.js:1
loadCamera @ artoolkit.min.js:1
ARCameraParam.load @ artoolkit.min.js:1
ARCameraParam @ artoolkit.min.js:1
(anonymous) @ arcontroller.component.js?version=0.64:129
(anonymous) @ artoolkit.min.js:1
Promise.then (async)
play @ artoolkit.min.js:1
success @ artoolkit.min.js:1
Promise.then (async)
ARController.getUserMedia @ artoolkit.min.js:1
ArControllerComponent.startAR @ arcontroller.component.js?version=0.64:125
trigger @ litegl.js?version=0.64:10617
Scene.start @ litescene.js?version=0.64:14400
changeState @ play.js?version=0.64:278
onPlay @ play.js?version=0.64:120
```

But it looks promising. Thank you

**Comment by kalwalt** [comment]
>  get a 404 when I run the NFT tracking:

Good to know, this not happens in localhost. I will fix this. At the moment does not load the NFT marker (probably for a wrong path) and as in AR.js two ARController are created one inside the WebWorker and another one outside (in the Component) . That is not optimal but we will see in the future for this problem.

**Comment by ThorstenBux** [comment]
Woohoo 🥳

**Comment by kalwalt** [comment]
> Woohoo

i haven't solved the 404 error but i made a little step ahead, and i have to found the correct method to pass the matrix transform to the renderer, maybe implementing an event listener, litescene.js should have his own system i think.

**Comment by ThorstenBux** [comment]
LiteScene does have its own event system yes. However js events should work too. Haven’t looked at your code yet. But can if you like.

Kind regards
Thorsten Bux

Sent from my iPhone

On 23/04/2020, at 10:07 AM, Walter Perdan <notifications@github.com> wrote:

﻿

Woohoo

i haven't solved the 404 error but i made a little step ahead, and i have to found the correct method to pass the matrix transform to the renderer, maybe implementing an event listener, litescene.js should have his own system i think.

**Comment by kalwalt** [comment]
I created the embedded worker but not solved the 404 issue, it should work, but it doesn't on the netlify services. Maybe is something related to the name of the base path? Because artooolkit.min.js should be at this path: `https://deploy-preview-34--sad-northcutt-73843d.netlify.app/editor/js/modules/artoolkit.min.js` but you get instead resource not available.

I also receive another error: 
```
Uncaught (in promise) ReferenceError: ar is not defined
    at process (webWorker.js:174)
    at startWorker (webWorker.js:198)
    at arcontroller.component.js?version=0.64:168
    at Array.forEach (<anonymous>)
    at ArControllerComponent.<anonymous> (arcontroller.component.js?version=0.64:158)
    at ARCameraParam.<anonymous> (artoolkit.min.js:1)
    at writeCallback (artoolkit.min.js:1)
    at writeByteArrayToFS (artoolkit.min.js:1)
    at XMLHttpRequest.oReq.onload (artoolkit.min.js:1)
```

**Comment by kalwalt** [comment]
Maybe `https://deploy-preview-34--sad-northcutt-73843d.netlify.app` start from the `editor` project folder?

**Comment by ThorstenBux** [comment]
Yes it does start from the editor folder

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by kalwalt** [comment]
@ThorstenBux with https://github.com/augmentmy-world/arStudio/pull/34/commits/952f94b6a0e6b9883790ed2b3a838da03be8aac8 fix the `404` error for the **importScripts**, now i have to fix the path for the camera and the marker.

**Comment by kalwalt** [comment]
Now **WebARStudio** can load `artoolkit.min.js`, `camera_para.dat` and the pinball **NFT** marker ! :tada: :rocket: :smile: Stopping for now, i will continue this evening.

**Comment by ThorstenBux** [comment]
Awesome, big shout out [Animated GIF]
________________________________

File: editor/js/modules/webWorker.js [hunk 1]
```diff
@@ -195,17 +194,31 @@ function startWorker(marker, video, input_width, input_height, canvas_draw) {
...
 
   var world;
 
-  var found = function(msg) {
+  var found = function (msg) {
     if (!msg) {
       world = null;
     } else {
       world = JSON.parse(msg.matrixGL_RH);
+      var senderM = new CustomEvent('sendMatrix', { detail: { world: world } });
```
Comment (kalwalt) [inline editor/js/modules/webWorker.js]: Probably i will need to change `detail` inside, should be conformant to data required by **onTrackableFound()**

File: editor/js/components/arcontroller.component.js [hunk 2]
```diff
@@ -153,7 +153,9 @@ ArControllerComponent.prototype.startAR = function() {
                 // Whenever ARController#process detects a marker, it fires a getMarker event
                 // with the marker details.
                 this.arController.addEventListener('getMarker',this.onTrackableFound.bind(this));
-                this.arController.addEventListener('getNFTMarker',this.onTrackableFound.bind(this));
+                document.addEventListener('sendMatrix', function (ev) {
```
Comment (kalwalt) [inline editor/js/components/arcontroller.component.js]: i attached the addEventListener to the document, it works but wondering if it is correct.

File: editor/js/modules/webWorker.js [hunk 3]
```diff
@@ -128,7 +128,7 @@ function startWorker(marker, video, input_width, input_height, canvas_draw) {
 
       importScripts(artoolkitUrl)
 
-      var onLoad = function () {
+      const onLoad = function () {
 
         ar = new ARController(msg.pw, msg.ph, param);
         var cameraMatrix = ar.getCameraMatrix();
```
Comment (kalwalt) [inline editor/js/modules/webWorker.js]: I don't know if we need cameraMatrix, i think we have that one from the ArControllerComponent, right?

**Comment by kalwalt** [comment]
@ThorstenBux i think we are arrived at a good point, just some refinements and will be super! Tested in localhost and the netlify deploy preview :smile: If you test look at the dev console while pointing to a pinball marker.

**Comment by kalwalt** [comment]
I have the listener now, but it output event in this form `ev.detail.data` instead onTrackableFound require  `event.data`. Not a clear idea how can i solve this problem, maybe i can try to implement for now a onTrackableNFTFound to skip this problem.

**Comment by kalwalt** [comment]
with the latest the NFT marker is tracked but i got this error:

```
artoolkit.min.js:1 [info] Tracked page 15794240 (max 0).
put_char @ artoolkit.min.js:1
write @ artoolkit.min.js:1
write @ artoolkit.min.js:1
doWritev @ artoolkit.min.js:1
_fd_write @ artoolkit.min.js:1
___wasi_fd_write @ artoolkit.min.js:1
jx @ artoolkit.min.js:10
$x @ artoolkit.min.js:10
jz @ artoolkit.min.js:10
iz @ artoolkit.min.js:10
Te @ artoolkit.min.js:6
Se @ artoolkit.min.js:6
dr @ artoolkit.min.js:12
Ws @ artoolkit.min.js:12
gP @ artoolkit.min.js:15
dynCall_iiii_32 @ VM28:4
getNFTMarker @ VM85:9
ARController.getNFTMarker @ artoolkit.min.js:1
ARController.process @ artoolkit.min.js:1
process @ cdf2cc15-8975-4de2-8582-890cf78bc384:97
self.onmessage @ cdf2cc15-8975-4de2-8582-890cf78bc384:12
arcontroller.component.js?version=0.64:357 saw a trackable with id 0
arcontroller.component.js?version=0.64:359 Uncaught ReferenceError: arTrackable is not defined
    at ArControllerComponent.onTrackableNFTFound (arcontroller.component.js?version=0.64:359)
    at found (webWorker.js:189)
    at Worker.worker.onmessage (webWorker.js:60)
ArControllerComponent.onTrackableNFTFound @ arcontroller.component.js?version=0.64:359
found @ webWorker.js:189
worker.onmessage @ webWorker.js:60
artoolkit.min.js:1 [info] Tracked page 15794240 (max 0).
```

**Comment by kalwalt** [comment]
This happens  obvioulsy because the addEventListener is attached to the window, not to the arController. This is one issue to be resolved.

**Comment by kalwalt** [comment]
I think i will close this PR better to start a new one with jsartoolkitNFT, or maybe i can start directly with webarStudio?

**Pull request closed without merging**